src/nvim/ex_docmd.c

```diff
@@ -1065,16 +1066,178 @@ void * getline_cookie(LineGetter fgetline,
 }
 
 /*
+ * Helper function to apply an offset for buffer commands, i.e. ":bdelete",
+ * ":bwipeout", etc.
+ * Returns the buffer number.
+ */
+static int compute_buffer_local_count(int addr_type, int lnum, int offset)
+{
+  buf_T *buf;
+  buf_T *nextbuf;
+  int count = offset;
+
+  buf = firstbuf;
+  while (buf->b_next != NULL && buf->b_fnum < lnum)
+    buf = buf->b_next;
+  while (count != 0) {
+    count += (count < 0) ? 1 : -1;
+    nextbuf = (offset < 0) ? buf->b_prev : buf->b_next;
+    if (nextbuf == NULL)
+      break;
+    buf = nextbuf;
+    if (addr_type == ADDR_LOADED_BUFFERS)
+      /* skip over unloaded buffers */
+      while (buf->b_ml.ml_mfp == NULL) {
+        nextbuf = (offset < 0) ? buf->b_prev : buf->b_next;
+        if (nextbuf == NULL) {
+          break;
+        }
+        buf = nextbuf;
+      }
+  }
+  // we might have gone too far, last buffer is not loaded
+  if (addr_type == ADDR_LOADED_BUFFERS) {
+    while (buf->b_ml.ml_mfp == NULL) {
+      nextbuf = (offset >= 0) ? buf->b_prev : buf->b_next;
+      if (nextbuf == NULL)
+        break;
+      buf = nextbuf;
+    }
+  }
+  return buf->b_fnum;
+}
+
+static int current_win_nr(win_T *win)
+{
+  int nr = 0;
+
+  FOR_ALL_WINDOWS_IN_TAB(wp, curtab) {
+    ++nr;
+    if (wp == win)
+      break;
+  }
+  return nr;
+}
+
+static int current_tab_nr(tabpage_T *tab)
+{
+  int nr = 0;
+
+  FOR_ALL_TABS(tp) {
+    ++nr;
+    if (tp == tab)
+      break;
+  }
+  return nr;
+}
+
+#define CURRENT_WIN_NR current_win_nr(curwin)
+#define LAST_WIN_NR current_win_nr(NULL)
+#define CURRENT_TAB_NR current_tab_nr(curtab)
+#define LAST_TAB_NR current_tab_nr(NULL)
+
+/*
+* Figure out the address type for ":wnncmd".
```

Comment:
Should say ":wincmd". Also might as well fix https://github.com/neovim/neovim/pull/2041/files#r28235543

Then after verifying the tests are working, let's merge this...